test/kad-dht.js

```diff
@@ -1,112 +1,545 @@
+        })
+      })
+
+      after(function (done) {
+        this.timeout(80 * 1000)
+        parallel(nodes.map((node) => (cb) => node.stop(cb)), done)
+      })
+
+      it('should get from the network after being added', function (done) {
+        this.timeout(60 * 1000)
+
+        addFileAndCat(nodes[1], [nodes[2], nodes[3]], data, done)
       })
     })
 
-    it('three hops', (done) => {
-      const data = crypto.randomBytes(9001)
+    describe('a Go bootstrap node in an hybrid land', () => {
+      let bootstrapAddr
+      let nodes = []
+      let data = crypto.randomBytes(9001)
+
+      // spawn bootstrap daemon and get address
+      before(function (done) {
+        this.timeout(60 * 1000)
+
+        spawnGoDaemon([], (err, node) => {
+          expect(err).to.not.exist()
+          nodes.push(node)
 
-      waterfall([
-        (cb) => goD3.api.add(data, cb),
-        (res, cb) => jsD.api.cat(res[0].hash, cb),
-        (stream, cb) => stream.pipe(bl(cb))
-      ], (err, file) => {
-        expect(err).to.not.exist()
-        expect(file).to.be.eql(data)
-        done()
+          getBootstrapAddr(nodes[0], (err, res) => {
+            expect(err).to.not.exist()
+            bootstrapAddr = res
+            done()
+          })
+        })
+      })
+
+      // spawn daemons
+      before(function (done) {
+        this.timeout(70 * 1000)
+        const bootstrap = [bootstrapAddr]
+
+        parallel([
+          (cb) => spawnJsDaemon(bootstrap, cb),
+          (cb) => spawnGoDaemon(bootstrap, cb),
+          (cb) => spawnJsDaemon(bootstrap, cb)
+        ], (err, spawnedNodes) => {
+          expect(err).to.not.exist()
+          spawnedNodes.forEach((node) => nodes.push(node))
+          setTimeout(done, 25000) // Wait for peers to have each other on the DHT
+        })
+      })
+
+      after(function (done) {
+        this.timeout(80 * 1000)
+        parallel(nodes.map((node) => (cb) => node.stop(cb)), done)
+      })
+
+      it('should get from the network after being added', function (done) {
+        this.timeout(60 * 1000)
+
+        addFileAndCat(nodes[1], [nodes[2], nodes[3]], data, done)
       })
     })
   })
 
-  describe('a Go node in the land of JS', () => {})
-  describe('hybrid', () => {})
+  describe('kad-dht with connections per peer', () => {
+    describe('a JS node in the land of Go', () => {
+      let nodes = []
+
+      // Spawn daemons
+      before(function (done) {
+        this.timeout(60 * 1000)
+
+        parallel([
+          (cb) => spawnGoDaemon([], cb),
+          (cb) => spawnGoDaemon([], cb),
+          (cb) => spawnGoDaemon([], cb),
+          (cb) => spawnJsDaemon([], cb)
+        ], (err, spawnedNodes) => {
+          expect(err).to.not.exist()
+          nodes = spawnedNodes
+          done()
+        })
+      })
+
+      // make connections
+      before(function (done) {
+        this.timeout(60 * 1000)
+
+        parallel(nodes.map((node) => (cb) => node.api.id(cb)), (err, ids) => {
```

Comment:
Suggestion: `async/each`

Reply:
In this case, I prefer this approach as using `each` would need to have an auxiliary array for keeping the ids for the second step.